contents/handbook/engineering/session-replay/session-replay-architecture.md

```diff
@@ -0,0 +1,217 @@
+   - Compresses each session block with Snappy
+4. **Flushes** periodically (max 10min buffer age or 50MB buffer size)
+
+**Persistence** (`sessions/s3-session-batch-writer.ts`):
+- Writes to **S3** as multipart uploads
+- File structure: `{prefix}/{timestamp}-{suffix}`
+- Each batch file contains multiple compressed session blocks
+- Uses byte-range URLs: `s3://bucket/key?range=bytes=start-end`
+- Retention-aware: writes to different S3 paths based on team retention period
+
+**Metadata** written to ClickHouse via Kafka:
+- Produces to **`clickhouse_session_replay_events`** topic
+- Table: **`session_replay_events`** (AggregatingMergeTree, sharded)
+- Stores: session_id, team_id, distinct_id, timestamps, URLs, counts (clicks/keypresses/console), block locations, retention_period_days
+- Old format also used: **`session_recording_events`** (deprecated, contains raw snapshot_data)
+
+## 3. Storage schema
+
+### ClickHouse tables
+**`session_replay_events`** (primary, v2):
+```
+session_id, team_id, distinct_id
+min_first_timestamp, max_last_timestamp
+block_first_timestamps[], block_last_timestamps[], block_urls[]
+first_url, all_urls[]
+click_count, keypress_count, mouse_activity_count, active_milliseconds
+console_log_count, console_warn_count, console_error_count
+size, message_count, event_count
+snapshot_source, snapshot_library
+retention_period_days
+```
+
+**`session_recording_events`** (legacy):
+- Stored raw `snapshot_data` directly in ClickHouse
+- Deprecated, being phased out
+
+### PostgreSQL
+**`posthog_sessionrecording`** model:
+- session_id (unique), team_id
+- object_storage_path (for LTS recordings)
+- full_recording_v2_path
+- Metadata: duration, active_seconds, click_count, start_time, end_time, distinct_id
+- Used for persisted/LTS recordings
+
+### S3 object storage
+- Main storage: `session_recordings/{team_id}/{session_id}/...`
+- Blob ingestion: organized by retention period + timestamp
+- Files are byte-addressable compressed session blocks
+
+## 4. Playback/Retrieval
+
+### API Flow (`posthog/session_recordings/session_recording_api.py`)
+
+**GET `/api/projects/:id/session_recordings/:session_id/`**:
+1. Loads metadata from ClickHouse `session_replay_events` or Postgres
+2. Returns: duration, start_time, person info, viewed status
+
+**GET `/api/projects/:id/session_recordings/:session_id/snapshots`**:
+Two-phase fetch:
+1. **Phase 1**: Returns available sources: `["blob"]` or `["blob", "realtime"]`
+    - note: "realtim"e" is deprecated and will be fully deleted soon
+2. **Phase 2**: Client requests `?source=blob`
+
+**Source resolution**:
+- **Blob (primary)**: Queries ClickHouse for block metadata
+  - Gets S3 URLs with byte ranges for each session block
+  - Generates pre-signed URLs (60s expiry)
+  - Client fetches compressed blocks directly from S3
+- **Legacy**: Old recordings stored in ClickHouse `session_recording_events` table
```

Comment:
definitely don't have any in CH now (they had 3 week TTL and we stopped >1.5 years ago :)))

useful history about the evolution but maybe can skip here